app.js

```diff
@@ -49,97 +50,144 @@ function isSafePositiveInteger(x) {
-    return;
-  }
-  if (opts && typeof opts.resume !== 'undefined' && typeof opts.resume !== 'boolean') {
-    cb(new error.BackupError('InvalidOption', 'Invalid resume option, must be type boolean'), null);
-    return;
-  }
-
   // Validate URL and ensure no auth if using key
   try {
     const urlObject = new URL(url);
     // We require a protocol, host and path (for db), fail if any is missing.
     if (urlObject.protocol !== 'https:' && urlObject.protocol !== 'http:') {
-      cb(new error.BackupError('InvalidOption', 'Invalid URL protocol.'));
-      return;
+      throw new OptionError('Invalid URL protocol.');
     }
     if (!urlObject.pathname || urlObject.pathname === '/') {
-      cb(new error.BackupError('InvalidOption', 'Invalid URL, missing path element (no database).'));
-      return;
+      throw new OptionError('Invalid URL, missing path element (no database).');
     }
-    if (opts && opts.iamApiKey && (urlObject.username || urlObject.password)) {
-      cb(new error.BackupError('InvalidOption', 'URL user information must not be supplied when using IAM API key.'));
-      return;
+    if (isIAM && (urlObject.username || urlObject.password)) {
+      throw new OptionError('URL user information must not be supplied when using IAM API key.');
     }
   } catch (err) {
-    cb(error.wrapPossibleInvalidUrlError(err));
-    return;
+    throw error.wrapPossibleInvalidUrlError(err);
   }
+  return true;
+}
 
+/**
+ * Validate options.
+ *
+ * @param {object} opts - Options.
+ * @returns Boolean true if all checks are passing.
+ */
+async function validateOptions(opts) {
+  // if we don't have opts then we'll be using defaults
+  if (!opts) {
+    return true;
+  }
+  const rules = [
+    { key: 'iamApiKey', type: 'string' },
+    { key: 'log', type: 'string' },
+    { key: 'output', type: 'string' },
+    { key: 'bufferSize', type: 'number' },
+    { key: 'parallelism', type: 'number' },
+    { key: 'requestTimeout', type: 'number' },
+    { key: 'mode', type: 'enum', values: ['full', 'shallow'] },
+    { key: 'resume', type: 'boolean' }
+  ];
+
+  for (const rule of rules) {
+    const val = opts[rule.key];
+    switch (rule.type) {
+      case 'string':
+        if (typeof val !== 'undefined' && typeof val !== 'string') {
+          throw new OptionError(`Invalid ${rule.key} option, must be type string`);
+        }
+        break;
+      case 'number':
+        if (typeof val !== 'undefined' && !isSafePositiveInteger(val)) {
+          throw new OptionError(`Invalid ${rule.key} option, must be a positive integer in the range (0, MAX_SAFE_INTEGER]`);
+        }
+        break;
+      case 'enum':
+        if (typeof val !== 'undefined' && rule.values.indexOf(val) === -1) {
+          // this is a shortcut. it should be join on rule.values,
+          // but we have just one enum, so hardcode will do for now
+          throw new OptionError('Invalid mode option, must be either "full" or "shallow"');
```

Comment:
Agreed, not really necessary when we have only one, but I think we can do this fairly easily:
```suggestion
          throw new OptionError(`Invalid mode option, must be one of: ${JSON.stringify(rule.values).slice(1,-1)`);
```

Reply:
This is missing "or" though, changing it for no-space comma separated. Something like `rule.values.map(x => JSON.stringify(x)).join(' or ')` is closer to what we have here, but that breaks on (future) 3 elements enum, so I just opted "we'll think about if we have to"

Reply:
Not _that_ simple, still doable. Done in https://github.com/IBM/couchbackup/pull/664/commits/d95662c3eed662c74d7cc2add47beb682fc1ef1d

Reply:
>This is missing "or" though

that's why I changed it to "one of", but anyway this is fine.